File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline

k, n = map(int, input().split())
li = []
for i in range(k):
    li.append(int(input()))
start = 0
end = max(li)

while start <= end:
    mid = (start + end) // 2
    if mid == 0:
        print(1)
        exit(0)
    cnt = 0
    for lan in li:
        cnt += lan // mid
    if cnt >= n:
        start = mid + 1
    else:
        end = mid - 1

print(end)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.swift
let kn = readLine()!.split(separator: " ").compactMap { Int($0) }
let (k, n) = (kn[0], kn[1])
var arr = [Int]()
for _ in 0..<k { arr.append(Int(readLine()!)!) }

var start = 1; var end = arr.max()!

while start <= end {
    let mid = (start + end) / 2
    let count = arr.reduce(0) { $0 + $1 / mid }

    if count >= n {
        start = mid + 1
    } else {
        end = mid - 1
    }
}

print(end)